dual_net.py

```diff
@@ -134,6 +139,45 @@ def should_log(i):
                     logger.add_summary(weight_summaries, global_step)
             self.save_weights()
 
+    def validate(self, tf_records, batch_size=128, init_from=None, num_steps=1000):
+        """Compute only the cost components for a set of tf_records, ideally a
+        holdout set, and report them to an 'eval' subdirectory of the logs.
+        """
+        cost_tensor_names = ['policy_cost', 'value_cost', 'l2_cost',
+                             'combined_cost']
+        if self.logdir is None:
+            print("Error, trainer not initialized with a logdir.")
+            return
+
+        logdir = self.logdir + '-eval'
+
+        with self.sess.graph.as_default():
+            input_tensors = preprocessing.get_input_tensors(
+                batch_size, tf_records, shuffle_buffer_size=10000, filter_amount=0.5)
+            output_tensors = dual_net(input_tensors, TRAIN_BATCH_SIZE,
+                                      train_mode=False, **self.hparams)
+            train_tensors = train_ops(
+                input_tensors, output_tensors, **self.hparams)
+
+            # just run our cost tensors
+            validate_tensors = {k: train_tensors[k]
+                                for k in cost_tensor_names + ['global_step']}
+            self.initialize_weights(init_from)
+            training_stats = StatisticsCollector()
+            logger = tf.summary.FileWriter(logdir, self.sess.graph)
+
+            for i in tqdm(range(num_steps)):
+                try:
+                    tensor_values = self.sess.run(validate_tensors)
+                except tf.errors.OutOfRangeError:
+                    break
+                training_stats.report(
```

Comment:
if you don't have global_step in here, you can just say `training_stats.report(tensor_values)`

Reply:
this makes more sense, will do.